drivers/counter/counter_sam_tc.c

```diff
@@ -0,0 +1,404 @@
+	}
+
+	if (chan_id == 0) {
+		tc_ch->TC_RA = alarm_value;
+		/* Clear interrupt status register */
+		(void)tc_ch->TC_SR;
+		tc_ch->TC_IER = TC_IER_CPAS;
+	} else {
+		tc_ch->TC_RB = alarm_value;
+		/* Clear interrupt status register */
+		(void)tc_ch->TC_SR;
+		tc_ch->TC_IER = TC_IER_CPBS;
+	}
+
+	LOG_DBG("set alarm: channel %u, count %u", chan_id, alarm_value);
+
+	return 0;
+}
+
+static int counter_sam_tc_cancel_alarm(const struct device *dev, uint8_t chan_id)
+{
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+
+	if (chan_id >= MAX_ALARMS_PER_TC_CHANNEL) {
+		return -EINVAL;
+	}
+
+	if (chan_id == 0) {
+		tc_ch->TC_IDR = TC_IDR_CPAS;
+		tc_ch->TC_RA = 0;
+	} else {
+		tc_ch->TC_IDR = TC_IDR_CPBS;
+		tc_ch->TC_RB = 0;
+	}
+
+	data->alarm[chan_id].callback = NULL;
+	data->alarm[chan_id].user_data = NULL;
+
+	LOG_DBG("cancel alarm: channel %u", chan_id);
+
+	return 0;
+}
+
+static int counter_sam_tc_set_top_value(const struct device *dev,
+					const struct counter_top_cfg *top_cfg)
+{
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+	int ret = 0;
+
+	for (int i = 0; i < MAX_ALARMS_PER_TC_CHANNEL; i++) {
+		if (data->alarm[i].callback) {
+			return -EBUSY;
+		}
+	}
+
+	/* Disable the compare interrupt */
+	tc_ch->TC_IDR = TC_IDR_CPCS;
+
+	data->top_cb = top_cfg->callback;
+	data->top_user_data = top_cfg->user_data;
+
+	tc_ch->TC_RC = top_cfg->ticks;
+
+	if ((top_cfg->flags & COUNTER_TOP_CFG_DONT_RESET) &&
+	    (tc_ch->TC_CV >= top_cfg->ticks)) {
+		ret = -ETIME;
+		if (top_cfg->flags & COUNTER_TOP_CFG_RESET_WHEN_LATE) {
+			tc_ch->TC_CCR = TC_CCR_SWTRG;
+		}
+	} else {
+		tc_ch->TC_CCR = TC_CCR_SWTRG;
+	}
+
+	/* Enable the compare interrupt */
+	tc_ch->TC_IER = TC_IER_CPCS;
+
+	return ret;
+}
+
+static uint32_t counter_sam_tc_get_top_value(const struct device *dev)
+{
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+
+	return tc_ch->TC_RC;
+}
+
+static uint32_t counter_sam_tc_get_pending_int(const struct device *dev)
+{
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+
+	return tc_ch->TC_SR & tc_ch->TC_IMR;
+}
+
+static void counter_sam_tc_isr(const struct device *dev)
+{
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+	uint32_t status;
+
+	status = tc_ch->TC_SR;
+
+	if (status & TC_SR_CPAS) {
+		if (data->alarm[0].callback) {
```

Comment:
shouldn't you somehow disable given interrupt so it does not get triggered on the next cycle?